### ISSUE TITLE
Java 6 Failure: NoSuchMethodError: java.net.URLClassLoader.close

### ISSUE THREAD

**Issue by djozis** [xtext/xtext-gradle-plugin#44]
Please let me know if bug reports for this Gradle plugin belong elsewhere - apologies if so.

Java: jdk1.6.0_45 (64bit)
Gradle: 2.10
Buildscript dependency: 'org.xtext:xtext-gradle-plugin:1.0.3'
Plugin: 'org.xtext.xtend'

:generateXtext step fails. Ran with -S for stacktrace and failure information:
```
FAILURE: Build failed with an exception.

* What went wrong:
Execution failed for task ':generateXtext'.
> java.net.URLClassLoader.close()V

* Try:
Run with --info or --debug option to get more log output.

* Exception is:
org.gradle.api.tasks.TaskExecutionException: Execution failed for task ':generateXtext'.
        at org.gradle.api.internal.tasks.execution.ExecuteActionsTaskExecuter.executeActions(ExecuteActionsTaskExecuter.java:69)
        at org.gradle.api.internal.tasks.execution.ExecuteActionsTaskExecuter.execute(ExecuteActionsTaskExecuter.java:46)
        at org.gradle.api.internal.tasks.execution.PostExecutionAnalysisTaskExecuter.execute(PostExecutionAnalysisTaskExecuter.java:35)
        at org.gradle.api.internal.tasks.execution.SkipUpToDateTaskExecuter.execute(SkipUpToDateTaskExecuter.java:64)
        at org.gradle.api.internal.tasks.execution.ValidatingTaskExecuter.execute(ValidatingTaskExecuter.java:58)
        at org.gradle.api.internal.tasks.execution.SkipEmptySourceFilesTaskExecuter.execute(SkipEmptySourceFilesTaskExecuter.java:52)
        at org.gradle.api.internal.tasks.execution.SkipTaskWithNoActionsExecuter.execute(SkipTaskWithNoActionsExecuter.java:52)
        at org.gradle.api.internal.tasks.execution.SkipOnlyIfTaskExecuter.execute(SkipOnlyIfTaskExecuter.java:53)
        at org.gradle.api.internal.tasks.execution.ExecuteAtMostOnceTaskExecuter.execute(ExecuteAtMostOnceTaskExecuter.java:43)
        at org.gradle.execution.taskgraph.DefaultTaskGraphExecuter$EventFiringTaskWorker.execute(DefaultTaskGraphExecuter.java:203)
        at org.gradle.execution.taskgraph.DefaultTaskGraphExecuter$EventFiringTaskWorker.execute(DefaultTaskGraphExecuter.java:185)
        at org.gradle.execution.taskgraph.AbstractTaskPlanExecutor$TaskExecutorWorker.processTask(AbstractTaskPlanExecutor.java:66)
        at org.gradle.execution.taskgraph.AbstractTaskPlanExecutor$TaskExecutorWorker.run(AbstractTaskPlanExecutor.java:50)
        at org.gradle.execution.taskgraph.DefaultTaskPlanExecutor.process(DefaultTaskPlanExecutor.java:25)
        at org.gradle.execution.taskgraph.DefaultTaskGraphExecuter.execute(DefaultTaskGraphExecuter.java:110)
        at org.gradle.execution.SelectedTaskExecutionAction.execute(SelectedTaskExecutionAction.java:37)
        at org.gradle.execution.DefaultBuildExecuter.execute(DefaultBuildExecuter.java:37)
        at org.gradle.execution.DefaultBuildExecuter.access$000(DefaultBuildExecuter.java:23)
        at org.gradle.execution.DefaultBuildExecuter$1.proceed(DefaultBuildExecuter.java:43)
        at org.gradle.execution.DryRunBuildExecutionAction.execute(DryRunBuildExecutionAction.java:32)
        at org.gradle.execution.DefaultBuildExecuter.execute(DefaultBuildExecuter.java:37)
        at org.gradle.execution.DefaultBuildExecuter.execute(DefaultBuildExecuter.java:30)
        at org.gradle.initialization.DefaultGradleLauncher$4.run(DefaultGradleLauncher.java:154)
        at org.gradle.internal.Factories$1.create(Factories.java:22)
        at org.gradle.internal.progress.DefaultBuildOperationExecutor.run(DefaultBuildOperationExecutor.java:90)
        at org.gradle.internal.progress.DefaultBuildOperationExecutor.run(DefaultBuildOperationExecutor.java:52)
        at org.gradle.initialization.DefaultGradleLauncher.doBuildStages(DefaultGradleLauncher.java:151)
        at org.gradle.initialization.DefaultGradleLauncher.access$200(DefaultGradleLauncher.java:32)
        at org.gradle.initialization.DefaultGradleLauncher$1.create(DefaultGradleLauncher.java:99)
        at org.gradle.initialization.DefaultGradleLauncher$1.create(DefaultGradleLauncher.java:93)
        at org.gradle.internal.progress.DefaultBuildOperationExecutor.run(DefaultBuildOperationExecutor.java:90)
        at org.gradle.internal.progress.DefaultBuildOperationExecutor.run(DefaultBuildOperationExecutor.java:62)
        at org.gradle.initialization.DefaultGradleLauncher.doBuild(DefaultGradleLauncher.java:93)
        at org.gradle.initialization.DefaultGradleLauncher.run(DefaultGradleLauncher.java:82)
        at org.gradle.launcher.exec.InProcessBuildActionExecuter$DefaultBuildController.run(InProcessBuildActionExecuter.java:94)
        at org.gradle.tooling.internal.provider.ExecuteBuildActionRunner.run(ExecuteBuildActionRunner.java:28)
        at org.gradle.launcher.exec.ChainingBuildActionRunner.run(ChainingBuildActionRunner.java:35)
        at org.gradle.launcher.exec.InProcessBuildActionExecuter.execute(InProcessBuildActionExecuter.java:43)
        at org.gradle.launcher.exec.InProcessBuildActionExecuter.execute(InProcessBuildActionExecuter.java:28)
        at org.gradle.launcher.exec.ContinuousBuildActionExecuter.execute(ContinuousBuildActionExecuter.java:78)
        at org.gradle.launcher.exec.ContinuousBuildActionExecuter.execute(ContinuousBuildActionExecuter.java:48)
        at org.gradle.launcher.exec.DaemonUsageSuggestingBuildActionExecuter.execute(DaemonUsageSuggestingBuildActionExecuter.java:51)
        at org.gradle.launcher.exec.DaemonUsageSuggestingBuildActionExecuter.execute(DaemonUsageSuggestingBuildActionExecuter.java:28)
        at org.gradle.launcher.cli.RunBuildAction.run(RunBuildAction.java:43)
        at org.gradle.internal.Actions$RunnableActionAdapter.execute(Actions.java:170)
        at org.gradle.launcher.cli.CommandLineActionFactory$ParseAndBuildAction.execute(CommandLineActionFactory.java:237)
        at org.gradle.launcher.cli.CommandLineActionFactory$ParseAndBuildAction.execute(CommandLineActionFactory.java:210)
        at org.gradle.launcher.cli.JavaRuntimeValidationAction.execute(JavaRuntimeValidationAction.java:35)
        at org.gradle.launcher.cli.JavaRuntimeValidationAction.execute(JavaRuntimeValidationAction.java:24)
        at org.gradle.launcher.cli.CommandLineActionFactory$WithLogging.execute(CommandLineActionFactory.java:206)
        at org.gradle.launcher.cli.CommandLineActionFactory$WithLogging.execute(CommandLineActionFactory.java:169)
        at org.gradle.launcher.cli.ExceptionReportingAction.execute(ExceptionReportingAction.java:33)
        at org.gradle.launcher.cli.ExceptionReportingAction.execute(ExceptionReportingAction.java:22)
        at org.gradle.launcher.Main.doAction(Main.java:33)
        at org.gradle.launcher.bootstrap.EntryPoint.run(EntryPoint.java:45)
        at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
        at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:39)
        at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:25)
        at java.lang.reflect.Method.invoke(Method.java:597)
        at org.gradle.launcher.bootstrap.ProcessBootstrap.runNoExit(ProcessBootstrap.java:54)
        at org.gradle.launcher.bootstrap.ProcessBootstrap.run(ProcessBootstrap.java:35)
        at org.gradle.launcher.GradleMain.main(GradleMain.java:23)
Caused by: java.lang.NoSuchMethodError: java.net.URLClassLoader.close()V
        at org.xtext.gradle.builder.XtextGradleBuilder.cleanup(XtextGradleBuilder.java:396)
        at org.xtext.gradle.builder.XtextGradleBuilder.doBuild(XtextGradleBuilder.java:347)
        at org.xtext.gradle.builder.XtextGradleBuilder.access$400(XtextGradleBuilder.java:69)
        at org.xtext.gradle.builder.XtextGradleBuilder$3.apply(XtextGradleBuilder.java:288)
        at org.xtext.gradle.builder.XtextGradleBuilder$3.apply(XtextGradleBuilder.java:251)
        at org.eclipse.xtext.xbase.lib.IteratorExtensions.forEach(IteratorExtensions.java:363)
        at org.eclipse.xtext.xbase.lib.IterableExtensions.forEach(IterableExtensions.java:333)
        at org.xtext.gradle.builder.XtextGradleBuilder.indexChangedClasspathEntries(XtextGradleBuilder.java:294)
        at org.xtext.gradle.builder.XtextGradleBuilder.build(XtextGradleBuilder.java:137)
        at org.xtext.gradle.tasks.XtextGenerate.build(XtextGenerate.java:344)
        at org.xtext.gradle.tasks.XtextGenerate.generate(XtextGenerate.java:209)
        at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
        at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:39)
        at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:25)
        at java.lang.reflect.Method.invoke(Method.java:597)
        at org.gradle.internal.reflect.JavaMethod.invoke(JavaMethod.java:75)
        at org.gradle.api.internal.project.taskfactory.AnnotationProcessingTaskFactory$IncrementalTaskAction.doExecute(AnnotationProcessingTaskFactory.java:244)
        at org.gradle.api.internal.project.taskfactory.AnnotationProcessingTaskFactory$StandardTaskAction.execute(AnnotationProcessingTaskFactory.java:220)
        at org.gradle.api.internal.project.taskfactory.AnnotationProcessingTaskFactory$IncrementalTaskAction.execute(AnnotationProcessingTaskFactory.java:231)
        at org.gradle.api.internal.project.taskfactory.AnnotationProcessingTaskFactory$StandardTaskAction.execute(AnnotationProcessingTaskFactory.java:209)
        at org.gradle.api.internal.tasks.execution.ExecuteActionsTaskExecuter.executeAction(ExecuteActionsTaskExecuter.java:80)
        at org.gradle.api.internal.tasks.execution.ExecuteActionsTaskExecuter.executeActions(ExecuteActionsTaskExecuter.java:61)
        ... 61 more
```
I'd like to add my thanks for what has been awesome software so far - I've been using it a lot and Google has solved all my issues so far except this. Is this plugin supposed to support Java 6?

**Comment by oehme** [comment]
This plugin currently needs at least Java 7. If it does not require too many changes, I would be happy for a contribution to make it work with Java 6. Would you like to look into that?

**Comment by djozis** [comment]
If time permits, it would be my pleasure. I'll post here if I get a chance to take a look at doing so.

In case someone else stumbles by this, I'll post a little more problem/workaround:
I was writing xtend code (compiled with Gradle) to be used by Java 6, originally compiling in Java 8 but setting:
```
sourceCompatibility = 1.6
targetCompatibility = 1.6
```
This didn't work because it used Java 8 classes like "Consumer" even though the bytecode was Java 6 compliant. Compiling with Java 6 didn't work at all, as above. While I can find no way to build with Gradle and have xtend compile against the Java 6 runtime libs/API, I found that Java 7 can run the plugin and didn't have the Java 8 API which was causing xtend generated Java to be non-compliant with the Java 6 API. Compiling with Java 7 targetting Java 6 compatability is my work-around for now.

Thanks for the fast response!

**Comment by oehme** [comment]
If you want to cross-compile against an older JDK, you can use the bootClasspath option. Xtext will pick it up from the JavaCompile task.

```
tasks.withType(JavaCompile) {
  options.fork = true
  options.bootClasspath = "path/to/jdk6"
}
```

**Comment by djozis** [comment]
Thanks for the tip - I was originally looking for that. Unfortunately, hard-coding the JDK location isn't really an option as other developers from my team need to be able to compile this too, and it needs to be able to build from Jenkins also.

**Comment by oehme** [comment]
You wouldn't hard-code it of course, you could either agree on an environment variable that everyone uses or have some JDK-detection logic that looks in the usual locations for a JDK 6 installation.

But if forcing the use of Java 7 works for you, that's cool too.

**Comment by kuniss** [comment]
We are putting developer specific Gradle variables (typically dev environment specific paths) to the gradle.properties file. In that way you have generic build scripts and developer specific builds. This is almost better than maintaining an environment variable (as it is OS agnostic).
E.g.
gradle.properties:
```
JDK6_HOME=C:/Program Files/java/jdk1.6.50_1
```
build.gradle:
```
tasks.withType(JavaCompile) {
  options.fork = true
  options.bootClasspath = "$JDK6_HOME"
}
```
This works also for Jenkins! 
The Jenkins system service's gradle.properties file is under C:\.gradle\gradle.properties if I remember right (... in case you are using a Windows machine).

**Comment by oehme** [comment]
@kuniss Yes that works too, although I would probably use a different file name like `local.properties`, because `gradle.properties` is meant to be checked in.

The user-scoped gradle properties are in <user home>/.gradle/gradle.properties

**Comment by kuniss** [comment]
@oehme Yes, I meant those user-scoped gradle properties file in ```<user home>/.gradle/gradle.properties```. :-)
Not the project's one.

**Comment by djozis** [comment]
Great advice. Perhaps it can be employed to make this plugin build using Java 6!

For my simple usage, I found no other conflicts aside from the line in the stacktrace above - skipping that line on Java 6 seems to be sufficient. https://github.com/xtext/xtext-gradle-plugin/pull/46

Ironically, now it builds on Java 7, targeting Java 6, but the Java 6 support is fragile because it doesn't target the Java 6 API (just like my work-around)... perhaps you can come up with a standard Java 6 lib dir setting for this project, as you've suggested.

**Comment by oehme** [comment]
Since we build on Travis CI, we can just set the build environment to JDK6 :)

**Comment by djozis** [comment]
Responded here: https://github.com/xtext/xtext-gradle-plugin/pull/46#issuecomment-200047633

**Comment by oehme** [comment]
Fixed by #46
